### DISCUSSION TITLE
iterators for arrays?

### DISCUSSION THREAD

**Discussion by pljakobs** [mikee47/ConfigDB#33] [Ideas]
I notice that in the Basic_Config example, you use `auto item = channels.addItem();` ,to iterate over an array, so auto type works for arrays. And of course, the code has to know the structure of the returned data object.
How difficult would it be to enable the use of `for (const auto item : channels) {}`?
I guess I can use `for (int i=0;i<channels.getItemCount(),i++)` but I wonder if the iterator style above can work as well?

 [edit - I think I messed this up in trying to copy stuff around and in doign so, made it unreadable. Glad you could figure out what I was trying to say :D]

**Comment by mikee47** [comment]
Yep, definitely want iterator support, just haven't got around to that yet. It's pretty easy since it'll just wrap the containing object and the index. I'll pop it on the pre-release issues list.

The `auto item = channels.addItem();` is currently equivalent to this:

```c++
BasicConfig::General::ContainedChannelsItem item = channels.addItem();
```

The IDE should be able to provide auto-completion though. There should be an alias so this works:

```c++
BasicConfig::General::Channels::Item item = channels.addItem();
```

I'll get that added.

**Reply by pljakobs** [reply]
excellent.. 
I've gotten about 50% of the rewrite of my codebase, the streaming input/output alone will save me more than 600LOC, making this so much easier to read.
I notice that it's a little bit of mental gymnastics to translate correctly between the schema and the actual implementation, especially since I have to keep in mind which object is a store and which isn't, but that's easy enough to get used to.

In order to conserve RAM, I am only ever keeping one store open, but there should not be an issue with having two or more open in the same database, right?

**Reply by mikee47** [reply]
> In order to conserve RAM, I am only ever keeping one store open, but there should not be an issue with having two or more open in the same database, right?

You can have as many stores as you like open at the same time from as many databases as you like, subject to memory availablity. I've included a `map` of the memory layout in the generated `.h` files which will give you an idea of how big each store is. Arrays (and strings) will obviously add to that but you should be able to get a pretty good estimate.
